src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs

```diff
@@ -0,0 +1,293 @@
+        }
+
+        /// <summary>
+        /// If true, the type is fully contained in the current compilation group.
+        /// </summary>
+        public override bool ContainsTypeLayout(TypeDesc type)
+        {
+            return _containsTypeLayoutCache.GetOrAdd(type, ContainsTypeLayoutUncached);
+        }
+
+        private bool ContainsTypeLayoutUncached(TypeDesc type)
+        {
+            if (type.IsObject ||
+                type.IsPrimitive ||
+                type.IsEnum ||
+                type.IsPointer ||
+                type.IsFunctionPointer ||
+                type.IsCanonicalDefinitionType(CanonicalFormKind.Any))
+            {
+                return true;
+            }
+            var defType = (MetadataType)type;
+            if (!ContainsType(defType))
+            {
+                if (!defType.IsValueType)
+                {
+                    // Eventually, we may respect the non-versionable attribute for reference types too. For now, we are going
+                    // to play is safe and ignore it.
+                    return false;
+                }
+
+                // Valuetypes with non-versionable attribute are candidates for fixed layout. Reject the rest.
+                if (!defType.HasCustomAttribute("System.Runtime.Versioning", "NonVersionableAttribute"))
+                {
+                    return false;
+                }
+            }
+            if (!defType.IsValueType && !ContainsTypeLayout(defType.BaseType))
+            {
+                return false;
+            }
+            foreach (FieldDesc field in defType.GetFields())
+            {
+                if (field.IsStatic)
+                    continue;
+
+                TypeDesc fieldType = field.FieldType;
+
+                if (fieldType.IsValueType &&
+                    !ContainsTypeLayout(fieldType))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public sealed override bool VersionsWithModule(ModuleDesc module)
+        {
+            return _versionBubbleModuleSet.Contains(module);
+        }
+
+        public sealed override bool VersionsWithType(TypeDesc typeDesc)
+        {
+            return typeDesc.GetTypeDefinition() is EcmaType ecmaType &&
+                _versionsWithTypeCache.GetOrAdd(typeDesc, ComputeTypeVersionsWithCode);
+        }
+
+
+        public sealed override bool VersionsWithMethodBody(MethodDesc method)
+        {
+            return _versionsWithMethodCache.GetOrAdd(method, VersionsWithMethodUncached);
+        }
+
+        private bool VersionsWithMethodUncached(MethodDesc method)
+        {
+            if (method.OwningType is MetadataType owningType)
+            {
+                if (!VersionsWithType(owningType))
+                    return false;
+            }
+            else
+                return false;
+
+            if (method == method.GetTypicalMethodDefinition())
```

Comment:
Should this be `GetMethodDefinition`? It doesn't seem like we need to take `ComputeInstantiationVersionsWithCode` below for non-generic methods on generic types.

Reply:
Yes that seems right. Functionally its the same, but calling the extra method is unnecessary.